pkg/storegateway/bucket.go

```diff
@@ -611,6 +611,11 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storegatewaypb.Stor
 	// but sometimes it can take minutes if the block isn't loaded and there is a surge in queries for unloaded blocks.
 	done, err := s.limitConcurrentQueries(ctx, stats)
 	if err != nil {
+		if errors.Is(err, errGateTimeout) {
```

Comment:
If a timeout occurs here, should we add an event to the trace span or log something?

It might also be useful to add a metric for the number of queries that were rejected due to the gate timing out - a high rate of this indicates either a performance problem or suggests scaling out store-gateways would be useful.

Reply:
> It might also be useful to add a metric for the number of queries that were rejected due to the gate timing out - a high rate of this indicates either a performance problem or suggests scaling out store-gateways would be useful.

A metric has been added directly in dskit:
https://github.com/grafana/dskit/pull/512

It doesn't tell exactly the reason why the gate wasn't allowed tho.

Reply:
> If a timeout occurs here, should we add an event to the trace span or log something?

good idea. I will add a span log

> It doesn't tell exactly the reason why the gate wasn't allowed tho.


with the current setup a query can be "not_permitted" either because it was cancelled (`context.Canceled`) or because it timed out (`context.DeadlineExceeded`). I can add this distinction in the metrics. There's also the response metrics of the gRPC method calls which would record `status_code="Canceled"` or `status_code="OK"`, but it might be more clear to have this separation in the dskit metrics